blackbird_python/blackbird/program.py

```diff
@@ -237,6 +255,37 @@ def __call__(self, **kwargs):
 
                     op['kwargs'][k] = func(**vals)
 
+        # set values for variables and arrays
+        for k, v in prog._var.items(): # pylint: disable=protected-access
+            # it can either be an independent parameter for a variable
+            if isinstance(v, sym.Expr):
+                par = list(v.free_symbols)
+                func = sym.lambdify(par, v)
+
+                try:
+                    vals = {str(p): kwargs[str(p)] for p in par}
+                except KeyError:
+                    raise ValueError("Invalid value for free parameter provided")
+
+                prog._var[k] = func(**vals)
+            # or encapsulated in an array
+            elif isinstance(v, np.ndarray):
+                populated_array = copy.deepcopy(v)
+                for i, row in enumerate(v):
+                    for j, col in enumerate(row):
+                        if isinstance(col, sym.Expr):
+                            par = list(col.free_symbols)
+                            func = sym.lambdify(par, col)
+
+                            try:
+                                vals = {str(p): kwargs[str(p)] for p in par}
+                            except KeyError:
+                                raise ValueError("Invalid value for free parameter provided")
+
+                            populated_array[i][j] = func(**vals)
```

Comment:
This code block (above as well) is particularly hard to parse!

Reply:
Added a comment here that briefly explains what's being done. :slightly_smiling_face: